### ISSUE TITLE
Questions about Reward in GSM8K

### ISSUE THREAD

**Issue by Xuekai-Zhu** [Yu-Fangxu/FoR#7]
Hi, this repository is truly excellent for exploring GFlowNet in reasoning tasks. I am currently delving into the implementation of the GSM8K task and noticed that there are some hard-coded elements that hinder me from reproducing your experiments.

I have a couple of questions:
**1.	How do you compute the reward for GSM8K?
2.	What is the OpenAI API used for?**

1. How do you compute the reward for GSM8K?
[FoR](https://github.com/Yu-Fangxu/FoR/tree/main)/[GSM8K](https://github.com/Yu-Fangxu/FoR/tree/main/GSM8K/lightning_module_selection.py) line 645-659.
From the following code, is there a binary reward for gsm8k ? 1e-4 or 100 ?

```
states.append(last_state)
        if len(answer_dict) == 0:
            r1 = 1e-4
            answer = None
        else:
            answer = sorted_answer_dict[0][0]
            if answer in goal:
                r1 = 100
            else:
                r1 = 1e-4


        r1 = torch.tensor(r1).to(self.device)
        intermediate_reward = torch.tensor(intermediate_reward)
        return intermediate_reward, actions, states, r1, answer
```

2.	What is the OpenAI API used for?
source code: from the following code, assuming you try to generate the response from `meta-llama/Llama-3-8B`. However, you set `self.client = OpenAI`, i can't reproduce these code.
```
world_output = self.query_LM(self.model, self.world_tokenizer, (world_input, f"{self.prompts['answer_prefix'].format(step+1)}"), do_sample=False, num_return_sequences=1,
                                    eos_token_id=eos_token_id, temperature=0)
```
```
self.client = OpenAI(
                    api_key=API_KEY,
                    base_url=API_BASE,
                    # timeout=300
                    )
```
```
def query_LM(self, worldmodel, tokenizer, prompt, eos_token_id, num_return_sequences=1, do_sample=True, temperature=1.0):
        all_results = []
        response = self.client.completions.create(
                   model='meta-llama/Llama-3-8B',
            prompt=prompt[0] + "\n" + prompt[1],
            n=num_return_sequences,
            max_tokens=150,
            temperature=temperature,
        )
        completions = response.choices

        for completion in completions:
            raw = completion.text
            raw = raw.split("\n")[0]
            result = prompt[1] + raw
            all_results.append(result)
        return all_results
```
3.  some hard-coded elements that hinder me from reproducing your experiments.
These paths might make it a bit challenging for me to reproduce the results. If you have time, perhaps you could consider changing them to relative paths.
**Once again, thank you for your contribution! This repo is a pioneer in applying GFlowNet to the field of reasoning!**
```
model.save_pretrained("/home/fangxu/GFlowPlan/ckpt/6-step")
        print("PEFT saved...")
        # trainer.test(model=task, datamodule=data)
        trainer.test(ckpt_path="last", dataloaders=data.test_dataloader())
        transition_path = f"/home/fangxu/GFlowPlan/transitions/{args.step}/transition.pkl"
        with open(transition_path, 'wb') as f:
            pickle.dump(task.transitions, f)
    else:
        model = load_model("meta-llama/Meta-Llama-3-8B", "./ckpt/6-step")
```

**Comment by Xuekai-Zhu** [comment]
**follow-up reminder**: sub-project of GSM8K missing a file `bw_utils.py`

**Comment by Xuekai-Zhu** [comment]
**Why did you use the learnable parameters to estimate Z ?**
Quoted from your paper:
<img width="969" alt="截屏2025-01-06 17 49 28" src="https://github.com/user-attachments/assets/2b7dfc76-d49e-45d0-b332-a300fdb7e225" />

However, your code is :
`logZ = torch.nn.Parameter(torch.tensor([args.logZ_init], dtype=torch.float))`

And as for more detail implementation:
```
loss = tb_loss(
            log_pf=LOG_PF,
            log_r=LOG_R_temperd,
            logz=self.logZ,
            log_bf=None,
            logpartition=True
        )
def tb_loss(log_pf, log_r, logz, log_bf=None, logpartition=False):
    print("log_pf: ", log_pf)
    print("log_r: ", log_r)
    print("logz: ", logz)
    if logpartition:
        if log_bf != None:
            scores = log_pf - log_r - log_bf
            loss = (scores - scores.mean()) ** 2 
        else:
            scores = log_pf - log_r
            loss = (scores - scores.mean()) ** 2 
    else:
        if log_bf != None:
            loss = (log_pf + logz - log_r - log_bf) ** 2
        else:
            loss = (log_pf + logz - log_r) ** 2
    return loss.mean()
```

**In summary, you use `scores = log_pf - log_r` and `loss = (scores - scores.mean()) ** 2` to obtain tb_loss, which is different with your Eq.5 in paper.** 

I am very curious about the implementation of GFlowNet in this work, as GFlowNet is indeed a relatively fundamental method, and applying it to complex tasks can be quite challenging. 

**Could you please share the reasoning behind this implementation? Specifically, the lessons and details behind the code design from the experiments? Thank you very much for your valuable contribution!**

**Comment by Yu-Fangxu** [comment]
Hi Xuekai,
Thanks for your attention to our work! I will respond to your questions one by one:

1. How do you compute the reward for GSM8K?
In the current version, I set the reward as success or not, which is a binary reward. I assign the reward as 100 when the reasoning reaches the correct answer, otherwise assign a small value, e.g. 1e-4. Extending GSM8K with a process reward like what we've done in BlocksWorld or training a process reward model is both feasible, but I have not yet implemented them.

2. What is the OpenAI API used for?
OpenAI API is used for vLLM to speed up the inference, you could refer to vLLM document https://docs.vllm.ai/en/latest/getting_started/examples/openai_chat_completion_client.html to find more details if you are not familiar with it.

3. follow-up reminder and some hard-coded elements that hinder me from reproducing your experiments.
Thanks! I remember I did not use bw_utils in GSM8K, I could further check it. And sorry for not cleaning the code, I will find some time to modify them!

4. Why did you use the learnable parameters to estimate Z?
The tb_loss is used when "logpartition=True", you could refer to line 240 in lightning_module_selection.py
You could also refer to another implementation in line 83:
https://github.com/GFNOrg/torchgfn/blob/master/src/gfn/gflownet/trajectory_balance.py

When using it in other tasks (e.g. molecular generation), one can choose the origianl trajectory balance.

5. Why logZ = torch.nn.Parameter(torch.tensor([args.logZ_init], dtype=torch.float))?
Thanks for pointing this out, I'm sorry for not cleaning the code, I do not use this variable actually. 
Some reasons for using eq6: When I started this project, I just wanted to learn a _logZ_ as proposed in the original GFlowNets. However, the total flow _Z_ is different for each data point. Therefore, learning a single value of _Z_ for all the data points does not make sense. One solution is adding an MLP to predict the value of _logZ_ given each data point and reasoning process, this may work theoretically but I'm not sure if it can accurately estimate _logZ_. To avoid explicitly assign the value for _Z_, I adopt the formulation in eq6, you could also check more details in **Zhang et al.**

**Comment by Xuekai-Zhu** [comment]
Mang thanks for your sharing !!!!
Following your suggestions, I successfully ran your code of GSM8K and GAME 24 on my device! 
And I have learned a lot from your lessons. 
To run the code of GSM8K task, i need to run `vllm serve Qwen/Qwen2.5-0.5B-Instruct` before, source tutorial in [vLLM Quickstart Guide](https://docs.vllm.ai/en/stable/getting_started/quickstart.html).

**Looking forward to future updates! I’ve shared your outstanding project with my friends and groups.**
Thank you once again for your detailed guidance and invaluable lessons!

**Comment by Yu-Fangxu** [comment]
Thank you Xuekai!

You may want to use llama-3 for better performance~

I think I should find some time to add more detailed instructions to allow users to run the code easily. Thanks again for your attention!
